mlos_bench/mlos_bench/storage/base_storage.py

```diff
@@ -0,0 +1,274 @@
+            """
+
+        @abstractmethod
+        def pending_trials(self) -> Iterator['Storage.Trial']:
+            """
+            Return an iterator over the pending trial runs for this experiment.
+            """
+
+        @abstractmethod
+        def new_trial(self, tunables: TunableGroups,
+                      config: Optional[Dict[str, Any]] = None) -> 'Storage.Trial':
+            """
+            Create a new experiment run in the storage.
+
+            Parameters
+            ----------
+            tunables : TunableGroups
+                Tunable parameters of the experiment.
+            config : dict
+                Key/value pairs of additional non-tunable parameters of the trial.
+
+            Returns
+            -------
+            trial : Storage.Trial
+                An object that allows to update the storage with
+                the results of the experiment trial run.
+            """
+
+    class Trial(metaclass=ABCMeta):
+        """
+        Base interface for storing the results of a single run of the experiment.
+        This class is instantiated in the `Storage.Experiment.trial()` method.
+        """
+
+        def __init__(self, *,
+                     tunables: TunableGroups, experiment_id: str, trial_id: int,
+                     config_id: int, opt_target: str, config: Optional[Dict[str, Any]] = None):
+            # pylint: disable=too-many-arguments
+            self._tunables = tunables
+            self._experiment_id = experiment_id
+            self._trial_id = trial_id
+            self._config_id = config_id
+            self._opt_target = opt_target
+            self._config = config or {}
+
+        def __repr__(self) -> str:
+            return f"{self._experiment_id}:{self._trial_id}"
+
+        @property
+        def trial_id(self) -> int:
+            """
+            ID of the current trial.
+            """
+            return self._trial_id
+
+        @property
+        def config_id(self) -> int:
+            """
+            ID of the current trial configuration.
+            """
+            return self._config_id
+
+        @property
+        def tunables(self) -> TunableGroups:
+            """
+            Tunable parameters of the current trial.
+            """
+            return self._tunables
+
+        def config(self, global_config: Optional[Dict[str, Any]] = None) -> dict:
+            """
+            Produce a copy of the global configuration updated
+            with the parameters of the current trial.
+            """
+            config = self._config.copy()
+            config.update(global_config or {})
+            config["experimentId"] = self._experiment_id
+            config["trialId"] = self._trial_id
+            return config
+
+        @abstractmethod
+        def update(self, status: Status,
+                   value: Optional[Union[Dict[str, Any], Any]] = None
+                   ) -> Optional[Dict[str, Any]]:
```

Comment:
Complete nit but we (myself included) are rather inconsistent with how we format function signatures with multiple lines.
In some the closing paren and return is on its own line like here, others (next function), they're grouped together with other params.
Some has multiple args per line, others one arg per line.
I don't care to fix this here necessarily, but maybe for the future we just pick one and stick to it.
I think my preference (probably borrowed from SQL) is either: 
- all on one line (if they fit), or
- one per line otherwise

Reply:
Same for function calls.

Reply:
I agree. I'll do it in another PR (I like to separate cosmetic changes from functionality updates)